src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -1019,41 +1020,83 @@ public static void JsonIgnoreConditionIsCorrectlyTranslatedToShouldSerializeDele
 
             static void TestJsonIgnoreConditionDelegate(JsonIgnoreCondition defaultIgnoreCondition, JsonIgnoreCondition? ignoreConditionOnProperty, JsonPropertyInfo property, ModifyJsonIgnore modify)
             {
-                // defaultIgnoreCondition is not taken into accound, we might expect null if defaultIgnoreCondition == ignoreConditionOnProperty
                 switch (ignoreConditionOnProperty)
                 {
-                    case null:
-                        Assert.Null(property.ShouldSerialize);
-                        break;
                     case JsonIgnoreCondition.Always:
                         Assert.NotNull(property.ShouldSerialize);
                         Assert.False(property.ShouldSerialize(null, null));
                         Assert.False(property.ShouldSerialize(null, ""));
                         Assert.False(property.ShouldSerialize(null, "asd"));
+                        Assert.Throws<InvalidCastException>(() => property.ShouldSerialize(null, 0));
```

Comment:
why invalid cast? shouldn't it always return `false`?

Reply:
I've updated the default delegates so that they are represented internally using `Func<object, T, bool>`. Results in less boxing for value properties but a side-effect of that is `InvalidCastException` when passing invalid arguments in the public, untyped delegate.